zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/http/ElasticsearchHttpSpanConsumer.java

```diff
@@ -48,16 +58,24 @@
       callback.onSuccess(null);
       return;
     }
+    Map<String, Set<Pair<String>>> indexToServiceSpans = new LinkedHashMap<>();
     try {
       HttpBulkIndexer indexer = new HttpBulkIndexer("index-span", es);
-      Map<String, Set<Pair<String>>> indexToServiceSpans = indexSpans(indexer, spans);
+      indexToServiceSpans = indexSpans(indexer, spans);
       if (!indexToServiceSpans.isEmpty()) {
         indexNames(indexer, indexToServiceSpans);
       }
       indexer.execute(callback);
     } catch (Throwable t) {
       propagateIfFatal(t);
       callback.onError(t);
+      synchronized(indexToServiceSpansCache) {
+          for (String index : indexToServiceSpans.keySet()) {
+              Set<Pair<String>> serviceSpansEntry = indexToServiceSpans.get(index);
+              Set<Pair<String>> serviceSpansEntryCached = indexToServiceSpansCache.get(index);
```

Comment:
looking at this, the logic is in a tight loop (for each span). I'd move the cache stuff to a higher scope, where you see `if (!indexToServiceSpans.isEmpty()) {`